readme.txt
PSXSDK 0.6.3
ReadMe and FAQ
Last updated: May 23rd 2021
-----------------------------

What is PSXSDK?

PSXSDK is an unofficial and homebrewed software development kit for the Sony PlayStation(TM).
Its programming tools (compiler, assembler, linker, etc.) are based on GNU Binutils and GNU
Compiler Collection (GCC) while the PlayStation-specific tools, which you need to convert data
to PSX format, were either written from scratch or adapted from already existing open source programs.

It is expected that you are already familiar with programming. 
If you want to learn how to program, I recommend starting programming for the operating system 
your computer uses and not here.

This readme is structured as a FAQ which also contains information which is not exclusive to the PSXSDK,
but that is generally informative.

Q: What do I need to run my programs on my PlayStation console?

A: There are several methods to do that, either a modchip, a boot disk, a cheating device
like Caetla, or the "swap trick".

The "modchip" method involves soldering a chip on the motherboard of your PlayStation to make it
think it's reading an original licensed disc. The kind of modchip depends on the model of your console,
and they are quite hard to find nowadays (as of 2010). You might already have one, as it was very
common to install it in many parts of the world.

I might be actually telling you to ignore your laws (in case you have EUCD and DMCA), and I recommend
to do it. The constant eroding of our rights as citizens is getting very annoying, and these
laws need to be fought. Let's not be hypocrites.
Big money influences law (TM).

The "boot disk" method uses a specially-made CD-ROM which looks like it is an original licensed disc
to the PlayStation. Once booted they show you a menu where you can boot an unlicensed disc.
You might be able to still find one for sale.

The "cheating device" method was very used back in the golden days of the PlayStation scene.
I have never used this method so I cannot describe it here. Search for "caetla" on your favourite
search engine and something interesting about that should come up.

The "swap trick" consists in first putting an original licensed disc in the PlayStation, so that
the PlayStation can authenticate, then when the drive is still spinning you replace it with an
unlicensed disc, and the PlayStation boots the disc as it already authenticated earlier.
The exact method of the "swap trick" varies between different models of the console.
It ruins your CD-ROM drive in the long run and spare PSX CD-ROM drives aren't very easy to find
(maybe harder to find than entire consoles), so I recommend to use something else if you can.
Search for this on your favourite search engine, something will surely come up.

On SCPH1001 PlayStations (Early American models) there seems to be an easier way to do the swap trick, also called audio menu exploit.
Make the SCPH1001 PlayStation go into the BIOS menu by not inserting any CD,
then go to the cd player, let it read an original PS1 disc until it stops spinning,
swap the disc with a burned CD-R disc, and exit the cd player.

This trick might also work on early Japanese consoles but I'm not sure of that.

Q: How do I get started?

A: In most cases, you will simply need one of the precompiled toolchains tarballs which you can find
on the PSXSDK website (http://unhaut.fav.cc/psxsdk). Extract them so that the PSXSDK can
be found at /usr/local/psxsdk and you're ready.

If you want to do it the hard way, there is no precompiled tarball for your operating system or
you want to help in developing the PSXSDK, read below on how to build it from sources.

Q: How do I compile a program?

A: This is easy. It is done almost as you do it usually, but with an additional step:

psx-gcc -o myprogram myprogram.c
elf2exe myprogram myprogram.exe

Why the additional step? It is needed because the GNU compiler outputs an ELF executable file,
which can't be run on the PlayStation. elf2exe converts that ELF file to a PS-X EXE, which
is the PlayStation's executable format, and which can be booted.

Q: How do I make a CDROM image?

A: I will first describe how to make your executable boot.
There are two ways to make your executable boot: you either name it "PSX.EXE" and you
put it in the CDROM filesystem's root directory, or you set a file named "SYSTEM.CNF" up
which describes a bit of information about your executable.

A SYSTEM.CNF file looks like this:

BOOT = cdrom:PROGRAM.EXE;1
TCB = 4
EVENT = 16
STACK = 801FFFF0

The BOOT field contains the filename of your program (in this case PROGRAM.EXE in the
root directory of the CDROM filesystem). I would advise you to not touch the other fields.
Modifying STACK is even useless (as the starting address of the stack is reset by the PSXSDK
library to 0x801FFFF0)

Now, you can make the ISO9660 filesystem for the CDROM with any program which does that.
This, for example, is the command you would use to do it with the cdrtools:

mkisofs -o test.iso -V TEST -sysid PLAYSTATION cd_root

Then you will have to use mkpsxiso to convert the ISO image to a format the PlayStation understands
and to license it. mkpsxiso outputs a .BIN/.CUE pair, which is supported by most CD burning
applications.
NOTE: Some Linux distributions (like Debian and Mint) don't provide mkisofs, but genisoimage instead.
           Replace the command name in the command line and that's it.

mkpsxiso test.iso test.bin /usr/local/psxsdk/misc/infousa.dat

This will create test.bin and test.cue, and the license applied to the image will be American.

Now you can burn the .BIN/.CUE pair with your favourite burning application.

Q: Where can I find the function reference for the PSXSDK?

A: You can find it either in /usr/local/psxsdk/misc/docs or in the docs subdirectory in the source
code archive.

Q: Are there any examples which show how to use the PSXSDK library?

A: Yes. Download the examples archive from the PSXSDK website and compile the examples in it.
In this way you can also test that your PSXSDK compiler is working correctly.
You can build them all by running ``make build_examples''

Q: What does PSXSDK support?

A: It can use the GPU (or GS, Graphic Synthesizer), the SPU, handle the VBlank Interrupt, handle
the joysticks and read files from the CDROM drive. Which is more than enough to do a 2D game.

Q: What is missing from PSXSDK?

A: Root counter support and a lot more...
Even if the PlayStation was very common back in its golden days, there is scarce documentation about it.
Since version 0.6.2 there is preliminary Geometry Transformation Engine (GTE) support provided by libmeidogte

Q: Why does the C library appear incomplete?

A: Many standard C library functions are wrapped around their implementation in the BIOS which keeps
code size low and eases implementation, but this also means that they have less features than
most C libraries of today. That isn't a big problem, because for games you don't need a lot of features.
Anyway, you can modify the PSXSDK to remove the wrapper and use your functions.
The PSXSDK already does that for BIOS functions which were found not working or too buggy (like memcpy).

Q: Help! My PlayStation crashed / something odd appeared on screen / other hardware bug

A: PSXSDK was primarily tested on a PAL SCPH-102 PSOne, on SCPH-5552, SCPH-7002 and SCPH-9002 PlayStations and on the PCSX emulator.
There are probably several hardware quirks and oddities I've not accounted for but which happen
on other models of the console. Please send an email if something like that happens.
Be sure to read realhw.txt in the documentation directory first, though, as it might be 
an easily fixable error on your part.

Q: What program can I use to convert my images to PlayStation format?

A: You can use bmp2tim which is included in the PSXSDK or any program which converts an image to
PlayStation TIM format. It is a fairly standard format in the PlayStation field.

Q: What program do I use in case I want to convert sound?

A: You can use the included wav2vag or any other program which converts to that format.
As it is the case for TIM, VAG is a quite standard format in the PlayStation environment.

VABs are just find compilations of more VAG files.

Q: Can I sell a game I make with my PSXSDK?

A: Yes, as you are not infringing any copyright.

Q: I want to use Caetla to upload and test programs on my PlayStation, but I cannot find a version of CatFlap for an operating system which is not DOS or Windows. Are there?

A: Very recently Fabio Battaglia (hkzlabs) released his CatFlap for Linux. You can use that on Linux, and use its source
code as a starting point for a port to other operating systems.

Q: When uploading and running programs with CatFlap, they hang after some PSXSDK initialization messages.
How can I make them not hang? I cannot run anything!

A: Insert a CD into the drive of your PlayStation. It doesn't have to be a PlayStation CD, any CD is good.
This happens due to the initialization routines of the CDROM subsystem. Hopefully it will be fixed someday.

Q: How can I change the writing and the logo at the "Licensed by Sony" PlayStation bootscreen?

A: That information is contained inside the license file that you use to license the CD-ROM image that you burn.
Thus, to modify the information you must use the "lictool" program included in the PSXSDK tools.
lictool modifies various information contained in the license file such as the writing, the logo, etc.
    
For example, to output a new modified American license file which uses your logo contained in new.tmd:
    
lictool /usr/local/psxsdk/share/licenses/infousa.dat my_new_license_file.dat -tmd=new.tmd
    
Remember to use my_new_license_file.dat when you use mkpsxiso, otherwise you will use the standard
American license file, and your new logo won't appear!
    
Logos are in TMD format, the format for 3D models in Sony's original development kit.
    
Q: My music module sounds wrong or weird when I play it with libmodplay

A: Most music module formats have a lot of built-in effects, which are often poorly documented,
very complex to implement and which are often thought for personal computer hardware which is less
limited than the PlayStation SPU in some aspects. Luckily most modules don't really use them
or sound quite right when those effects are not implemented.

Q: I heard that it is possible to make the PlayStation read CD-RW discs (CD-rewritable). Is this true?

A: Yes, it is. But it is only really usable if you want to do hardware programming tests, as the amount
that the CD drive of your PlayStation may be able to read might be around 128-150 kilobytes. You will
not be able to read original PlayStation games or CD-Rs, and if you try to do it, your laser pickup might
get stuck and you may have to unblock it by unscrewing it and sliding it with your hands...
With a relatively good laser and very good tuning you might be even able to play full fledged games.

To read CD-RWs on later model PlayStation/cd drives you have to tune the potentiometer on the CD drive
from its original position to 12-13 hours. It actually sort of looks like a clock. Remember the original
position to not ruin anything. Use a precision screwdriver to tune the potentiometer.
This will stress the laser in the CD drive more than it is in normal operation, and it will shorten its life.

The trick to tuning the potentiometer correctly is burning an audio track on the CD-RW and playing the audio track
in the PlayStation CD player. If the audio track plays smoothly enough, it will be able to read CD-RWs.

You may need several tries before you succeed booting a disk, this happens due to the extensive read checks done
by the PlayStation BIOS when booting a game.

This is my (Giuseppe Gatta's) experience, and it might work much better on your PlayStation/PS cd drive
or it might work much worse. PlayStation cd drives are very finicky, I myself have PlayStations which read
like new and some which almost do not read at all. The CD drive is the lowest quality item inside
a PlayStation.

On the other hand this is the cheapest way to test things on your modded PlayStation.

Q: I don't use double buffering in my program/game. Nothing is shown on the emulator but it displays fine
    on the real hardware. Where is the issue?
    
A: You are using a buggy GPU plugin. If you're using version 1.0.16 of the P.E.Op.S SoftX Driver or most other
P.E.Op.S GPU plugins, update to P.E.Op.S SoftX Driver 1.0.18 and the problem should go away.

Q: What is a .PSF music file?

A: A .PSF music file is in the Portable Sound Format which is used to store video game music data.
Actually the .PSF just contains a tiny bit of information data in its header and then a compressed executable
which can be executed by an emulator (or a video game console). A PSF player is just a stripped-down emulator,
which supports only the SPU and the executable inside the PSF is an executable hacked to play only music.
Originally it was meant only for the PlayStation but now there are music files in this platform-indipendent format
from many consoles. You can play a .psf music just by using your console or emulator. Use the psfex tool shipped
with the PSXSDK, extract the .exe from the PSF and then run the executable. 
More info here: http://www.neillcorlett.com/psf/

Q: Why are filenames limited to the MSDOS-like 8.3 naming convention?

A: Because the ISO-9660 filesystem used on PlayStation CDs has the 8.3 limit for filenames.
It is plain ISO-9660 without any extension like Joliet or Rock Ridge, which are commonly added on CDs meant
to be read on computers to overcome the filename limit and to add new features.
The PSXSDK can make use of TRANS.TBL files in the directories of your CD to emulate long filenames.
TRANS.TBL is a normal text file which assigns long filenames to normal ISO9660 filenames.
Its rows look like this:
F SYSTEM.CNF;1		system.cnf

What this row does is assign to long filename system.cnf the short filename SYSTEM.CNF;1
'F' denotes that system.cnf is a file, otherwise 'D' denotes directories.

Good ISO image making tools will support the automatic creation of a TRANS.TBL file for each directory at image creation.
For instance, adding the -T parameter to your mkisofs command line makes mkisofs do just that.

To make fopen() use long filenames you will have to use the cdromL: pseudo device instead of cdrom:
Spaces are allowed in paths and both the slash and the back slash can be used to separate directories.
Examples: cdromL:/Path To/My Filename , cdromL:\Path To\This Name

There is a perfomance penalty when using long filenames because doing that requires reading the
TRANS.TBL for every directory which is encountered.

firstfile() does not support long filenames (yet) and open() does not support them, either.

Q: I'm running <linux distribution> and audio doesn't work in PCSX!
A: Many new Linux distributions are using PulseAudio. Run 'padsp pcsx' instead of 'pcsx'.

Q: Why is this SDK called PSXSDK? The PSX is a DVR+PS2 hybrid released by Sony in Japan!
A: Almost nobody knows about the Japanese PSX from 2004, and on top of that, it was a marketing flop.
    Seriously, PSX ('PlayStation eXperimental') is the name the original PlayStation went by during      development and it continued being called like that later. The name stuck and it is here to stay.

Q: Can I use UPX to pack an executable I made with PSXSDK?
A: Sure, you can. Due to a technicality of elf2exe, you need to specify --force on the UPX command line.
    It seems to work well, but I did no extensive testing, so it is at your own risk.

Q: Why do programs not work on the ePSXe/pSX/... emulators?
A: Despite popular belief, ePSXe and pSX are not as accurate as it is thought.
    It is recommended to use PCSX Reloaded for testing.
    The rule of thumb is that when something works fine on real hardware, and the
    emulator freezes or does something unexpected, it's the emulator's fault.

Written by Giuseppe Gatta
E-mail: tails92@gmail.com


Changelog
2021-05-23 (0.6.3):
- Updated this version of PSXSDK to function on Mac OSX (10.15.7); multiple file amended in tools/
- Added gitignore.

2019-04-10 (0.6.2):
- Included GTE library by LameGuy64/Meido-Tek as 'libmeidogte'
- New example: gtecube. Demonstrates libmeidogte
- The 'rottest' and 'puzzle' examples always set a PAL video mode regardless of the setting in Makefile.cfg
  Fixed.
- Removed unused macro code from spasm

2018-01-15 (0.6.1) :
- Fixed bug where LO and HI registers were not saved and restored by the NoBIOS exception handler
- Some definitions (like the ones for the interrupt controller and the GPU) moved out of
library source code to public headers, so they are available to all applications that may need them
- Fixes in pad.c so that compiler optimizations do not generate code with bad behavior
- Fixed crash when using a GCC 7.x+ compiler when calling static constructors during initialization.
- Static constructors are now called in backwards order. (as they are on most systems)
- Fixed bug where the __ctor_list symbol wouldn't have the address of the constructor list,
but would have its address offset unpredictably (leading to the crash mentioned above).
- Integrated upstream changes from F3M in libf3m
- Removed all references to Composer 669 music format in the sources and 
  documentation of libmodplay, as support for the format was never 
  implemented.
- Added psxs3m example. Mostly the same as psxmod but using libf3m 
instead of libmodplay
- Fixed some issues in the psxmod example.

2016-06-03 (0.6):
- size_t and ssize_t definitions moved from stdio.h to types.h, and types.h is included automatically by stdio.h
- Added strcoll() - as a wrapper to strcmp()
- File device enum is now public and moved to stdio.h 
  It was incorrectly private and in libpsx/src/libc.c before
- fopen(): path argument is now const char*
- rename() and remove(): both arguments are now const char*
- open(): filename argument is now const char*
- get_real_file_size() and get_file_size(): argument is now const char*
- firstfile(): name argument is now const char*
- Added stdin, stdout and stderr FILE structures
- Implemented fwrite(). Only works with stdout and stderr
- Implemented fputs(). Only works with stdout and stderr
- Implemented freopen()
- Implemented clearerr(), feof(), ferror() and fileno()
- Implemented placeholders for strerror() and strerror_r()
- Added libf3m, to playback S3M module files, adapted to PSXSDK from GreaseMonkey's F3M
- libm/k_rem_pio2.c: fixed misleading indentation at line 188.
  libm/e_asin.c: fixed misleading indentation at line 89.
  libm/w_exp.c: o_threshold and u_threshold variables only defined when
  _IEEE_LIBM is not defined
  Fixed warnings with GCC 6.1.0
- Compilation warnings fixed for some tools, and some cleanup in spasm/parser.c
- Fixed warnings with psxgte.h - specifier was put after and not before enumeration
- Fixed warnings in the examples

2015-07-29 (0.5.99):
- Added PSX_WarmReboot()
- Fixed extern directive for GsCurDrawEnvH: was wrongly specified for GsCurDrawEnvY
- NORMAL_LUMINOSITY is now an alias for NORMAL_LUMINANCE: "luminosity" is an incorrect term for the meaning
- Added PRFONT_RL(f), PRFONT_GL(f), RLFONT_BL(f) attributes to GsSetFontAttrib(). These allow to adjust the luminance factor for the color components used when drawing the font: in few words, when used correctly, the attributes can be used to change the color of the font on the fly, without using another color look-up table.
- Added set_cop_register(), get_cop_register() and run_cop_instruction().
  set_cop0_register() and get_cop0_register() are now aliases for set_cop_register() and
  get_cop_register() respectively.
- Added set_cop_ctrl_register() and get_cop_ctrl_register()
- Fixed some bugs in the Makefiles, thanks to rsoft and Missingno_force for providing patches
- Fixed a bug in bmp2tim where 4-bit images with a width not a multiply of 4, or 8-bit images with a width
  not a multiply of 8 were erronously generated and thus contained invalid data.
  It should not have been possible to generate such images in the first place.
- Fixed a bug in bmp2tim where unnecessary debug messages were printed on screen when converting an image.
- The path for objcopy hardcoded inside elf2exe is now taken from the OBJCOPY variable specified in Makefile.cfg,
  and not from the OBJCOPY_PATH define in the source code itself
- Added -fno-strict-overflow flag by default to Makefile.cfg and to the psx-gcc and psx-g++ scripts
- Multiplication registers HI and LO are now saved and restored by the vblank and root counter handlers -
  this could give problems if any division or multiplication instruction was used inside a handler.
  Fortunately no such instructions are used inside handlers most times.
- Greatly improved nv-SPASM - now it should be fully compatible (maybe save some quirks) 
  with the Hitmen assembler
- Added GsVPrintFont()

2014-04-22 (0.4.99):
- Fixed the Makefile for the tools, now it considers the HOST_* variables and removes executables
  correctly under Windows
- added --enable-languages=c to the command line to build GCC, in toolchain.txt
- Added abort() to libc
- Fixed Makefile for misc directory
- Fixed genscript.sh: generated link script playstation.x now correctly
  includes libgcc and does not rely on EXTERN directives
- New SsUpload() implementation contributed by Shendo, which contains workarounds for emulators. 
  Also works on ePSXe.
- Fixed wrong return value for McWriteSector()
- Introduced debug build, that can be built by specifying the PSXSDK_DEBUG
  flag at compile time - it has more verbose messages, diagnostics, etc.
- Very verbose messages are now only compiled in for the debug builds.
- Added support for the 1KB scratchpad, which is the data cache that went
  unused in the PlayStation - it acts as a kind of "fast RAM";

test/window/window.c
#include <stdio.h>
#include <psx.h>
#include <psxgpu.h>

// Program variables
/* Screen resolution */
#define SCREEN_XRES		320
#define SCREEN_YRES		240

#define OT_LEN 0x4000 /* 128 kb */

unsigned int prim_list[OT_LEN];
volatile int vblank_happened = 0;

// Current frame buffer indicator
int current_buf = 0;

GsDrawEnv game_drawenv;
GsDispEnv game_dispenv;

GsRectangle game_rect;


// Program functions
/* Dont really understand this function and variable yet... */
static void program_vblank_handler(void) {
	vblank_happened = 1;
}

/* Initialise game environmental variables */ 
void init(void)
{
    /* Initialize PSXSDK Library */
	PSX_InitEx(PSX_INIT_CD);
	/* Initialize the Graphic Synthesizer */
	GsInit();
	/* Clear video memory */
	GsClearMem();

	// Set up drawing environment
	game_drawenv.dither = 0;
	game_drawenv.draw_on_display = 0;
	game_drawenv.x = 0;
	game_drawenv.y = 0;
	game_drawenv.w = SCREEN_XRES;
	game_drawenv.h = SCREEN_YRES;
	game_drawenv.ignore_mask = 0;
	game_drawenv.set_mask = 0;

	GsSetDrawEnv(&game_drawenv);
	
	// Set up display environment
	game_dispenv.x = 0;
	game_dispenv.y = 256;
	
	GsSetDispEnv(&game_dispenv);

    /* Set primitive list pointer */
	GsSetList(prim_list);

    /* Set 320x240 video mode */
	GsSetVideoMode(320, 240, VMODE_NTSC);

	/* Set VBlank Handler */
	SetVBlankHandler( program_vblank_handler );	
}

// MAIN DISPLAY OR DRAW FUNCTION.
// This is where we draw everything
void display(void)
{
    /* Change display and drawing environment settings */
	/* for double buffering */
	if (current_buf) 
	{
		game_drawenv.y = 0;
		GsSetDrawEnv(&game_drawenv);
		game_dispenv.y = 256;	
		GsSetDispEnv(&game_dispenv);
	}
	else 
	{
		game_drawenv.y = 256;
		GsSetDrawEnv(&game_drawenv);
		game_dispenv.y = 0;
		GsSetDispEnv(&game_dispenv);
	}
	current_buf = !current_buf;
	
	
	/* Clear the screen with the specified color */
	GsSortCls(34, 53, 226);

	// Set perameters for window, or rect 1.
	game_rect.x = 10;
	game_rect.y = 20;
	game_rect.w = 300;
	game_rect.h = 200;
	game_rect.r = 0;
	game_rect.g = 0;
	game_rect.b = 0;
	game_rect.attribute = 0;
	
	// Bind parameters to drawing list
	GsSortRectangle(&game_rect);

	// Establish new window or rectangle parameters
	game_rect.x = 20;
	game_rect.y = 40;
	game_rect.w = 280;
	game_rect.h = 160;
	game_rect.r = 255;
	game_rect.g = 255;
	game_rect.b = 255;
	game_rect.attribute = 0;
	
	// Bind parameters to drawing list
	GsSortRectangle(&game_rect);

    /* Draw list */
    GsDrawList();
}

int main(void)
{
    init();

    while (1) 
    {
        display();
    }
    return 0;
}


examples/gtecube/gtecube.c
/* 
 * GTE Graphics Example (Spinning Cube)
 * 2019 Meido-Tek Productions
 * 2019 modified by nextvolume for PSXSDK
 *
 * licensed under Mozilla Public License v2.0 
 */
 
#include <stdio.h>
#include <string.h>
#include <psx.h>
#include <psxgpu.h>
#include <meidogte.h>

/* Screen resolution */
#define SCREEN_XRES		320
#define SCREEN_YRES		240

/* Screen center position */
#define CENTERX			SCREEN_XRES>>1
#define CENTERY			SCREEN_YRES>>1

#define OT_LEN 0x8000

/* For easier handling of vertex indices */
typedef struct {
	short v0,v1,v2,v3;
} INDEX;

/* Cube vertices */
SVECTOR cube_verts[] = {
	{ -100, -100, -100, 0 },
	{  100, -100, -100, 0 },
	{ -100,  100, -100, 0 },
	{  100,  100, -100, 0 },
	{  100, -100,  100, 0 },
	{ -100, -100,  100, 0 },
	{  100,  100,  100, 0 },
	{ -100,  100,  100, 0 }
};

/* Cube face normals */
SVECTOR cube_norms[] = {
	{ 0, 0, -ONE, 0 },
	{ 0, 0, ONE, 0 },
	{ 0, -ONE, 0, 0 },
	{ 0, ONE, 0, 0 },
	{ -ONE, 0, 0, 0 },
	{ ONE, 0, 0, 0 }
};

/* Cube vertex indices */
INDEX cube_indices[] = {
	{ 0, 1, 2, 3 },
	{ 4, 5, 6, 7 },
	{ 5, 4, 0, 1 },
	{ 6, 7, 3, 2 },
	{ 0, 2, 5, 7 },
	{ 3, 1, 6, 4 }
};

/* Number of faces of cube */
#define CUBE_FACES 6

/* Light color matrix */
/* Each column represents the color matrix of each light source and is */
/* used as material color when using gte_ncs() or multiplied by a */
/* source color when using gte_nccs(). 4096 is 1.0 in this matrix */
/* A column of zeroes disables the light source. */
MATRIX color_mtx = { .m = {
	{ONE, 0, 0},	/* Red   */
	{ONE, 0, 0},	/* Green */
	{ONE, 0, 0}	/* Blue  */ }
};

/* Light matrix */
/* Each row represents a vector direction of each light source. */
/* An entire row of zeroes disables the light source. */
MATRIX light_mtx = { .m = {
	/* X,  Y,  Z */
	{ -2048 , -2048 , -2048},
	{0	  , 0	  , 0},
	{0	  , 0	  , 0} }
};


/* Function declarations */
void init(void);
void display(void);

unsigned int prim_list[OT_LEN];
int current_buf = 0;
volatile int vblank_happened = 0;

// GsPoly4 is part of psxgpu.h 
GsPoly4 poly[CUBE_FACES];					/* Flat shaded quad primitive */

static void program_vblank_handler(void) {
	vblank_happened = 1;
}

enum {
	OP_SPIN, 
	OP_COLORR1, OP_COLORG1, OP_COLORB1, 
	OP_COLORR2, OP_COLORG2, OP_COLORB2, 
	OP_COLORR3, OP_COLORG3, OP_COLORB3, 
	OP_LIGHTX1, OP_LIGHTY1, OP_LIGHTZ1, 
	OP_LIGHTX2, OP_LIGHTY2, OP_LIGHTZ2, 
	OP_LIGHTX3, OP_LIGHTY3, OP_LIGHTZ3, 
	OP_SCRXOFF, OP_SCRYOFF,
	OP_SCRDEPTH,
	OP_TRANSVX, OP_TRANSVY, OP_TRANSVZ,
	OP_END
};

const char *operationNames[] = {"Change spinning speed",
	"Change light color R1", "Change light color G1", "Change light color B1", 
	"Change light color R2", "Change light color G2", "Change light color B2", 
	"Change light color R3", "Change light color G3", "Change light color B3", 
	"Change light X1", "Change light Y1", "Change light Z1", 
	"Change light X2", "Change light Y2", "Change light Z2", 
	"Change light X3", "Change light Y3", "Change light Z3", 
	"Change screen X offset", "Change screen Y offset",
	"Change screen depth",
	"Change translation vector X", "Change translation vector Y", 
	"Change translation vector Z", 
	};

int stopped = 0;
int currentOperation = 0;

int spinningSpeed = 16;
int screenXOffset = CENTERX;
int screenYOffset = CENTERY;
int screenDepth = CENTERX;
	
VECTOR	pos = { .vx = 0, .vy = 0, .vz = 400 };	/* Translation vector for TransMatrix */
	
#define COLOR_STEP	32
#define LIGHT_STEP		128

/* Main function */
int main() {

	int i,p,wasStart=0,wasLeft=0,wasRight=0;
	int a,b;
	short coord[2];
	unsigned char rgb[3] = {255, 255, 0};
	unsigned short padbuf;
	
	SVECTOR	rot;			/* Rotation vector for Rotmatrix */
	MATRIX	mtx,lmtx;				/* Rotation matrices for geometry and lighting */
	
	bzero(&rot, sizeof(rot));
	
	/* Init graphics and GTE */
	init();

	/* Main loop */
	while( 1 ) {
		/* Set GTE offset (recommended method  of centering) */
		gte_SetGeomOffset( screenXOffset, screenYOffset );
	
		/* Set screen depth (basically FOV control, W/2 works best) */
		gte_SetGeomScreen(screenDepth);
		
		/* Set light ambient color and light color matrix */
		gte_SetBackColor( 63, 63, 63 );
		gte_SetColorMatrix( &color_mtx );
		
		/* Set rotation and translation to the matrix */
		RotMatrix( &rot, &mtx );
		TransMatrix( &mtx, &pos );
		
		/* Multiply light matrix by rotation matrix so light source */
		/* won't appear relative to the model's rotation */
		MulMatrix0( &light_mtx, &mtx, &lmtx );
		
		/* Set rotation and translation matrix */
		gte_SetRotMatrix( &mtx );
		gte_SetTransMatrix( &mtx );
		
		/* Set light matrix */
		gte_SetLightMatrix( &lmtx );
		
		if (!stopped) {
		/* Make the cube SPEEN */
			rot.vx += spinningSpeed;
			rot.vz += spinningSpeed;
		}
			
		for( i=0; i<CUBE_FACES; i++ ) {
			
			/* Load the first 3 vertices of a quad to the GTE */
			gte_ldv3( 
				&cube_verts[cube_indices[i].v0], 
				&cube_verts[cube_indices[i].v1], 
				&cube_verts[cube_indices[i].v2] );
				
			/* Rotation, Translation and Perspective Triple */
			gte_rtpt();
			
			/* Compute normal clip for backface culling */
			gte_nclip();
			
			/* Get result*/
			gte_stopz( &p );
			
			/* Skip this face if backfaced */
			if( p < 0 ) {
				poly[i].attribute = -1;
				continue;
			}
			
			/* Calculate average Z for depth sorting */
			gte_avsz4();
			gte_stotz( &p );
			
			/* Skip if clipping off */
			/* (the shift right operator is to scale the depth precision) */
			if( (p>>2) > OT_LEN ) {
				poly[i].attribute = -1;
				continue;
			}
			
			/* Set the projected vertices to the primitive */
			gte_stsxy0( &coord );
			poly[i].x[0] = coord[0];
			poly[i].y[0] = coord[1];
			
			gte_stsxy1( &coord );
			poly[i].x[1] = coord[0];
			poly[i].y[1] = coord[1];
			
			gte_stsxy2( &coord );
			poly[i].x[2] = coord[0];
			poly[i].y[2] = coord[1];
			
			/* Compute the last vertex and set the result */
			gte_ldv0( &cube_verts[cube_indices[i].v3] );
			gte_rtps();
			gte_stsxy( &coord );
			poly[i].x[3] = coord[0];
			poly[i].y[3] = coord[1];			

			rgb[0] = 255;
			rgb[1] = 255;
			rgb[2] = 255;
			
			gte_ldrgb( &rgb );
			
			/* Load the face normal */
			gte_ldv0( &cube_norms[i] );
			
			/* Normal Color Single */
			gte_nccs();
			
			/* Store result to the primitive */
			gte_strgb( &rgb );
			
			poly[i].r = rgb[0];
			poly[i].g = rgb[1];
			poly[i].b = rgb[2];

			poly[i].attribute = 0;
		}
		
		/* Swap buffers and draw the primitives */
		display();
		
		/* Wait for VBlank to happen */
		vblank_happened = 0;
		while(!vblank_happened);
		
		/* Read joypad status */
		PSX_ReadPad(&padbuf, NULL);
		
		/* Logic for real-time 'configuration' of the cube */
		if ((padbuf & PAD_START) && !wasStart) {
			stopped = !stopped;
			wasStart = 1;
		}
		
		if ((padbuf & PAD_LEFT) && !wasLeft) {
			currentOperation--;
			
			if (currentOperation < 0)
				currentOperation = OP_END - 1;
			
			wasLeft = 1;
		}
		
		if ((padbuf & PAD_RIGHT) && !wasRight) {
			currentOperation++;
			
			if (currentOperation >= OP_END)
				currentOperation = 0;
			
			wasRight = 1;
		}
		
		if (padbuf & PAD_UP) {
			switch(currentOperation) {
				case OP_SPIN:
					spinningSpeed++;
				break;
				case OP_COLORR1 ... OP_COLORB3:
					a = (currentOperation - OP_COLORR1);
					b = a % 3;
					a /= 3;
				
					if(color_mtx.m[b][a] < ONE) 
						color_mtx.m[b][a]+=COLOR_STEP;
				break;
				case OP_LIGHTX1 ... OP_LIGHTZ3:
					a = (currentOperation - OP_LIGHTX1);
					b = a % 3;
					a /= 3;
				
					if(light_mtx.m[a][b] < ONE) 
						light_mtx.m[a][b]+=LIGHT_STEP;
				break;
				case OP_SCRXOFF:
					screenXOffset++;
				break;
				case OP_SCRYOFF:
					screenYOffset++;
				break;
				case OP_SCRDEPTH:
					screenDepth++;
				break;
				case OP_TRANSVX:
					pos.vx++;
				break;
				case OP_TRANSVY:
					pos.vy++;
				break;
				case OP_TRANSVZ:
					pos.vz++;
				break;
			}
		}
		
		if (padbuf & PAD_DOWN) {
			switch(currentOperation) {
				case OP_SPIN:
					spinningSpeed--;
				break;
				case OP_COLORR1 ... OP_COLORB3:
					a = (currentOperation - OP_COLORR1);
					b = a % 3;
					a /= 3;
				
					if(color_mtx.m[b][a] > 0) 
						color_mtx.m[b][a]-=COLOR_STEP;
				break;
				case OP_LIGHTX1 ... OP_LIGHTZ3:
					a = (currentOperation - OP_LIGHTX1);
					b = a % 3;
					a /= 3;
				
					if(light_mtx.m[a][b] > 0) 
						light_mtx.m[a][b]-=LIGHT_STEP;
				break;
				case OP_SCRXOFF:
					screenXOffset--;
				break;
				case OP_SCRYOFF:
					screenYOffset--;
				break;
				case OP_SCRDEPTH:
					screenDepth--;
				break;
				case OP_TRANSVX:
					pos.vx--;
				break;
				case OP_TRANSVY:
					pos.vy--;
				break;
				case OP_TRANSVZ:
					pos.vz--;
				break;
			}
		}
		
		if (!(padbuf & PAD_START))
			wasStart = 0;
		
		if (!(padbuf & PAD_LEFT))
			wasLeft = 0;
		
		if (!(padbuf & PAD_RIGHT))
			wasRight = 0;
	}
	
	return 0;
}

void init(void) {
	/* Initialize PSXSDK */
	PSX_InitEx( 0 );
	/* Initialize the Graphic Synthesizer */
	GsInit();
	/* Clear video memory */
	GsClearMem();
	/* Set primitive list pointer */
	GsSetList( prim_list );
	/* Set automatic GPU waiting */
	GsSetAutoWait();
	/* Set 320x240 video mode */
	GsSetVideoMode(320, 240, EXAMPLES_VMODE);
	/* Set VBlank Handler */
	SetVBlankHandler( program_vblank_handler );
	
	/* Load text font in video memory */
	GsLoadFont(768, 0, 768, 256);

	/* Initialize the MeidoGTE library */
	InitGeom();
}

void display(void) {
	int i;
	
	/* Change display and drawing environment settings */
	/* for double buffering */
	GsSetDispEnvSimple(0, current_buf ? 0 : 256);
	GsSetDrawEnvSimple(0, current_buf ? 256 : 0, SCREEN_XRES, SCREEN_YRES);
	current_buf = !current_buf;
	
	/* Clear the screen with the specified color */
	GsSortCls(63, 0, 127);
	
	/* Add cube faces to drawing list */
	for(i=0;i<CUBE_FACES;i++) {
		if(poly[i].attribute != -1)
			GsSortPoly4(&poly[i]);
	}
	
	/* If stopped, show so on the screen */
	if (stopped)
		GsPrintFont(0, 0, "Stopped");
	
	/* Print current operation name on screen */
	GsPrintFont(0, GsScreenH-16, "%s", operationNames[currentOperation]);
	
	/* Draw list */
	GsDrawList();
}
